sys/aarch64/rootdev.c

```diff
@@ -45,11 +77,52 @@ static bus_space_t *rootdev_bus_space = &(bus_space_t){
 
 static void rootdev_intr_setup(device_t *dev, unsigned num,
                                intr_handler_t *handler) {
-  /* TODO(pj) */
+  rootdev_t *rd = dev->parent->state;
+
+  assert(num < N_IRQ);
+  intr_event_t *event = &rd->intr_event[num];
+  intr_event_add_handler(event, handler);
 }
 
 static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
-  /* TODO(pj) */
+  intr_event_remove_handler(handler);
+}
+
+/* Read 32 bit pending register located at va and run handlers. */
+static inline void intr_handle(vaddr_t va, intr_event_t *events) {
+  assert(va != 0);
+  assert(events != NULL);
+
+  uint32_t pending = *(uint32_t *)va;
+
+  for (uint32_t irq = 0; irq < 32; irq++) {
```

Comment:
Umh... in most cases pending will be either 0 or will have a single bit. Could you optimize for such cases. Please use `ffs`.

Reply:
Do you have to manually apply interrupt mask?